main.tf

```diff
@@ -14,100 +14,162 @@
+      max_connections_per_instance = lookup(backend.value, "max_connections_per_instance", null)
+      max_rate                     = lookup(backend.value, "max_rate", null)
+      max_rate_per_instance        = lookup(backend.value, "max_rate_per_instance", null)
+      max_utilization              = lookup(backend.value, "max_utilization", null)
+    }
+  }
+
+  health_checks = [
+    google_compute_health_check.default[tonumber(each.value)].self_link]
+
+  security_policy = var.security_policy
+  enable_cdn      = var.cdn
 }
 
-resource "google_compute_http_health_check" "default" {
-  project      = "${var.project}"
-  count        = "${length(var.backend_params)}"
-  name         = "${var.name}-backend-${count.index}"
-  request_path = "${element(split(",", element(var.backend_params, count.index)), 0)}"
-  port         = "${element(split(",", element(var.backend_params, count.index)), 2)}"
+resource "google_compute_health_check" "default" {
+  # Note:
+  #  - tf supports only sets of stings, so options below don't work
+  #     for_each = local.bp_maps # - not supported type
+  #     for_each = toset(range(length(local.bp_maps))) # - not supported type
+  #     for_each = toset(keys(var.backends)) # requires calling type conversion function
+  #  - keeping the old interface(var.backed_params) is a bit too expensive
+  # without changing the itnerface count.index works better than for_each
+  count = length(local.bp_maps)
+  name = "${var.name}-${local.bp_maps[count.index]["port_name"]}-${local.bp_maps[count.index]["port"]}-${local.bp_maps[count.index]["substr_hashed_path"]}"
+  timeout_sec = local.bp_maps[count.index]["timeout"]
+  check_interval_sec = local.bp_maps[count.index]["timeout"] + local.bp_maps[count.index]["timeout"] / 2
+  project     = var.project
+
+  http_health_check {
+    request_path = local.bp_maps[count.index]["path"]
+    port         = local.bp_maps[count.index]["port"]
+  }
 }
 
-# Create firewall rule for each backend in each network specified, uses mod behavior of element().
-resource "google_compute_firewall" "default-hc" {
-  count         = "${length(var.firewall_networks) * length(var.backend_params)}"
-  project       = "${element(var.firewall_projects, count.index) == "default" ? var.project : element(var.firewall_projects, count.index)}"
-  name          = "${var.name}-hc-${count.index}"
-  network       = "${element(var.firewall_networks, count.index)}"
-  source_ranges = ["130.211.0.0/22", "35.191.0.0/16", "209.85.152.0/22", "209.85.204.0/22"]
-  target_tags   = ["${var.target_tags}"]
-
-  allow {
-    protocol = "tcp"
-    ports    = ["${element(split(",", element(split("|", join("", list(join("|", var.backend_params), replace(format("%*s", length(var.backend_params), ""), " ", "|")))), count.index)), 2)}"]
+# Create firewall rule for each backend in each network specified
+resource "google_compute_firewall" "default" {
+  for_each = toset(var.firewall_networks)
```

Comment:
Why have you removed backend_params?

Reply:
It fails. Instead we create a resource per network and then we iterate over backend_params to create dynamic allow block for each firewall rule.